src/com/sap/piper/DefaultValueCache.groovy

```diff
@@ -51,13 +55,56 @@ class DefaultValueCache implements Serializable {
             if(customDefaults in List)
                 configFileList += customDefaults
             for (def configFileName : configFileList){
-                if(configFileList.size() > 1) steps.echo "Loading configuration file '${configFileName}'"
-                def configuration = steps.readYaml text: steps.libraryResource(configFileName)
+                if(configFileList.size() > 1) script.echo "Loading configuration file '${configFileName}'"
+                def configuration = script.readYaml text: script.libraryResource(configFileName)
                 defaultValues = MapUtils.merge(
                         MapUtils.pruneNulls(defaultValues),
                         MapUtils.pruneNulls(configuration))
             }
             DefaultValueCache.createInstance(defaultValues, customDefaults)
+
+            persistDefaults(script, defaultValues, customDefaults)
+
+        }
+    }
+
+    static def persistDefaults(Script script, Map defaultValues, List customDefaults = []) {
+        if (!script) {
+            return null
```

Comment:
Unfortunately,  [commonPipelineEnvironment](https://github.com/o-liver/jenkins-library/blob/e2cd785dc4eb5f9406b23984560fe71af7303c50/vars/commonPipelineEnvironment.groovy#L139) consumes the defaults passing `null` as script.

Reply:
The default value cache is persisted with the `prepareDefaultValues` step, which is called [here in `setupCommonPipelineEnvironment`](https://github.com/SAP/jenkins-library/blob/master/vars/setupCommonPipelineEnvironment.groovy#L38).
To restore the default value cache after restart we need to provide a script to the `getInstance` method, which is called by the [`ConfigurationLoader`](https://github.com/SAP/jenkins-library/pull/1123/files#diff-3e6d1283a08d9fe02397e113f94de2d7), which is called by the [`ConfigurationHelper`](https://github.com/SAP/jenkins-library/pull/1123/files#diff-df89dc11d0df356cc9033325deffc242), which is used by each step.

**Now**, should we also fix the mentioned issue above which is affecting the steps [`piperInitRunStageConfiguration`](https://github.com/SAP/jenkins-library/blob/master/vars/piperInitRunStageConfiguration.groovy#L64), [`piperPipeline`](https://github.com/SAP/jenkins-library/blob/master/vars/piperPipeline.groovy) and [`piperPipelineStagePost`](https://github.com/SAP/jenkins-library/blob/master/vars/piperPipelineStagePost.groovy#L45);
**Or**, should we leave this PR as it is and fix the above issue with another PR?
The consequence if we choose the latter would be that the bug on restart remains present for these steps.

Reply:
My fear is it will increase the size of this PR too much and it will not be so easily reviewed, that is why I would like to fix these steps in a separate PR.

Reply:
The commonPipelineEnvironment  is not supposed to consider a script handle. I'd say we keep it as it is, so your PR will improve the code but wont fix the "restart" issue.  
With the Go implementation we probably we re-design the configuration handling.

Reply:
All this PR is intended to fix is the "restart" issue for the default values. Do you disagree that this is fixed with this PR @rodibrin ?

Reply:
Yes, even for the default values only i guess the restart feature isn't save. The code is quite complex, so i can't be sure, but in some places the DefaultValueCache is accessed without script handle (script=null).  If the instance is null, then no default values will be returned. E.g. see [commonPipelineEnvironment](https://github.com/o-liver/jenkins-library/blob/d40a6481101153bcf615ac4187b570b680472f02/vars/commonPipelineEnvironment.groovy#L149)